FastText/src/models/custom_layers.jl

```diff
@@ -0,0 +1,300 @@
+    return cell
+end
+
+function (m::WeightDroppedLSTMCell)((h, c), x)
+    b, o = m.b, size(h, 1)
+    Wi = Flux._isactive(m) ? m.Wi .* m.maskWi : m.Wi
+    Wh = Flux._isactive(m) ? m.Wh .* m.maskWh : m.Wh
+    g = Wi*x .+ Wh*h .+ b
+    input = σ.(gate(g, o, 1))
+    forget = σ.(gate(g, o, 2))
+    cell = tanh.(gate(g, o, 3))
+    output = σ.(gate(g, o, 4))
+    c = forget .* c .+ input .* cell
+    h′ = output .* tanh.(c)
+    return (h′, c), h′
+end
+
+Flux.@functor WeightDroppedLSTMCell
+
+Flux.trainable(m::WeightDroppedLSTMCell) = (m.Wi, m.Wh, m.b, m.h, m.c)
+
+testmode!(m::WeightDroppedLSTMCell, mode=true) =
+  (m.active = (isnothing(mode) || mode == :auto) ? nothing : !mode; m)
+
+"""
+    WeightDroppedLSTM(in::Integer, out::Integer, p::Float64=0.0)
+
+WeightDroppedLSTM layer
+
+This makes the WeightDroppedLSTMCell stateful, by wrapping the layer in a Recur field.
+
+Defining an instance:
+
+julia> wd = WeightDroppedLSTM(4, 5, 0.3);
+"""
+function WeightDroppedLSTM(a...; kw...)
+    cell = WeightDroppedLSTMCell(a...;kw...)
+    maskWi = drop_mask(cell.Wi, cell.p)
+    maskWh = drop_mask(cell.Wh, cell.p)
+    hidden = (cell.state0..., maskWi, maskWh)
+    return Flux.Recur(cell, hidden)
+end
+
+"""
+    reset!(m)
+
+Resets the h, c parameters of the LSTM Cell.
+
+For more refer [`Flux.reset`](@ref https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.reset!)
+"""
+function reset!(layers)
+    for layer in layers
+        if typeof(layer) == FastText.WeightDroppedLSTM
+            (layer.layer.state = (layer.layer.cell.h, layer.layer.cell.c))
+        else
+            Flux.reset!(layer)
+        end
+    end
+end
+
+
+"""
+    reset_masks!(layer)
+
+This is an important funciton since it used to reset the masks
+which are saved in WeightDroppedLSTMCell after every pass.
+
+julia> wd = WeightDroppedLSTM()
+
+julia> reset_masks!(wd)
+"""
+function Flux.reset!(layer::Flux.Recur{<:WeightDroppedLSTMCell})
+    maskWi = drop_mask(layer.cell.Wi, layer.cell.p)
+    maskWh = drop_mask(layer.cell.Wh, layer.cell.p)
+    layer.state = (layer.cell.state0..., maskWi, maskWh)
+    
+    return nothing
+end
+####################################################################
+
+########################## Varitional DropOut ######################
+"""
+    VarDrop(p::Float64=0.0)
+
+Variational Dropout layer
+
+Unlike standard dropout layer, which applies new dropout mask to every Array ot Matrix passed to it,
+this layer saves the mask applied till it is explicitly set to reset mode using 'reset_masks!'.
+
+Usage:
+julia> vd = VarDrop()
+
+To reset mask:
+julia> reset_masks!(vd)
+"""
+
+mutable struct VarDrop{F}
+    p::F
+    mask
+    reset::Bool
+    active::Bool
+end
+
+VarDrop(p::Float64=0.0) = VarDrop(p, Array{Float32, 2}(UndefInitializer(), 0, 0), true, true)
+
+function (vd::VarDrop)(x)
+    vd.active || return x
+    if vd.reset
+        vd.mask = drop_mask(x, vd.p)
+        vd.reset = false
+    end
+    return (x .* vd.mask)
+end
```

Comment:
As promised, here's the proposal from our call today:
```julia
mutable struct VarDrop{F}
    p::F
    active::Union{Bool, Nothing} # matches other norm layers
end

VarDrop(p::Real=0.0) = VarDrop(p, nothing)

function (vd::VarDrop)((has_mask, mask), x)
    if active
        mask = has_mask ? mask : drop_mask(x, vd.p)
        return (true, mask), x .* mask
    elseif !has_mask
        return (has_mask, mask), x
    else
        error("Mask set but layer is in test mode. Call `reset!` to clear the mask.")
    end
end
```
Plus bits to make `testmode!`, `_isactive` and `reset!` work properly.